### PR TITLE
Updates for class

### PR DESCRIPTION
**Pull request by vgoliber**
Program is restructured to use unit tests to score student solutions with autograder (requires inset functions).

### CODE REVIEW

File: npp.py [hunk 1]
```diff
@@ -15,23 +15,41 @@
 ## ------- Set up our list of numbers -------
 S = [25, 7, 13, 31, 42, 17, 21, 10]
 
+# TODO: Enter your token here
+token = 'Your-Token-Here'
```
Comment (m3ller) [inline npp.py]: Is there a particular reason why we are inserting in the token directly when it can be set up in dwave-config?

File: npp.py [hunk 2]
```diff
@@ -15,23 +15,41 @@
...
+
+Q = define_QUBO(S)
 
 ## ------- Run our QUBO on the QPU -------
-# TODO:  Choose QPU parameters
-chainstrength = 1
-numruns = 10
 
 # Run the QUBO on the solver from your config file
 from dwave.system import DWaveSampler, EmbeddingComposite
```
Comment (m3ller) [inline npp.py]: I know that this is not part of the PR, but is there a reason why the import line is here? Generally, they are at the top of the page.

pep8: https://www.python.org/dev/peps/pep-0008/#imports

File: npp.py [hunk 3]
```diff
@@ -15,23 +15,41 @@
...
+sampler = EmbeddingComposite(DWaveSampler(endpoint='https://cloud.dwavesys.com/sapi/', token=token, solver={'qpu': True}))
+
+def run_on_QPU(Q, sampler):
```
Comment (m3ller) [inline npp.py]: What is the purpose of this function?

If the user is expected to tune the QPU parameters, then `run_on_QPU(..)` is really:
```
def run_on_QPU(Q, sampler, chainstrength,  numruns):
    sampleset = sampler.sample_qubo(Q, chain_strength=chainstrength, num_reads=numruns)
    return sampleset
```

Since it's essentially wrapping `sample_qubo(..)` and not executing anything else, I think this function is better off as just part of the the main executable script.

File: npp.py [hunk 4]
```diff
@@ -15,23 +15,41 @@
 ## ------- Set up our list of numbers -------
 S = [25, 7, 13, 31, 42, 17, 21, 10]
 
+# TODO: Enter your token here
+token = 'Your-Token-Here'
+
 ## ------- Set up our QUBO dictionary -------
 
 # TODO:  Add code here to define your QUBO dictionary
+def define_QUBO(S):
```
Comment (m3ller) [inline npp.py]: Based on context, I think this function is meant to grab a number-partitioning QUBO for a set `S`. Hence, the function isn't allowing the end-user to define a QUBO, it's grabbing a very specific QUBO for the end-user.

Perhaps it can be written as
```
# TODO: Add code here ...
def get_qubo(S):
    ```Returns a number partitioning QUBO for a list of numbers S```

    Q = {}
    ..
```

File: npp.py (hunk 1 above)
Comment (vgoliber) [reply npp.py]: the autograder on our grading platform (Gradescope) does not have a config file

File: npp.py (hunk 3 above)
Comment (vgoliber) [reply npp.py]: the function is there so that the autograder can score their results - they have to be returned from a function to set up the unit test to grade it.

File: npp.py (hunk 4 above)
Comment (vgoliber) [reply npp.py]: The user/student is defining the QUBO within that code block.  Please keep in mind this is written for a very specific class exercise with a very specific autograder with limited functionality.

File: npp.py [hunk 5]
```diff
@@ -12,26 +12,50 @@
...
 
-## ------- Set up our list of numbers -------
-S = [25, 7, 13, 31, 42, 17, 21, 10]
-
-## ------- Set up our QUBO dictionary -------
+## ------- import packages -------
+from dwave.system import DWaveSampler, EmbeddingComposite
 
 # TODO:  Add code here to define your QUBO dictionary
+def define_QUBO(S):
```
Comment (m3ller) [inline npp.py]: Perhaps just a docstring here? That way a student will know right away what type of data structure `S` is.

Generally, functions are all in snake-case. (i.e. lower case and separated by underscores). https://www.python.org/dev/peps/pep-0008/#function-and-variable-names

```
def define_qubo(S):
    ```Returns a dictionary representing a QUBO.

    Args:
        S(list of integers): represents the numbers being partitioned
    ```
    Q = {}
```

File: npp.py [hunk 6]
```diff
@@ -12,26 +12,50 @@
...
+    Q = {}
 
-# Run the QUBO on the solver from your config file
-from dwave.system import DWaveSampler, EmbeddingComposite
+    # Add QUBO construction here
+    
+    return Q
+
+# TODO:  Choose QPU parameters in the following function
+def run_on_QPU(Q, sampler):
```
Comment (m3ller) [inline npp.py]: ditto on docstring and on snake-case

```
def run_on_qpu(Q, sampler):
    ```What this function does.

    Args:
        Q(dict): a representation of a QUBO
        sampler(dimod.Sampler): a sampler that uses the QPU
    ```
```

**Review by m3ller — APPROVED** [review]
looks good! just some small suggestions

File: npp.py [hunk 7]
```diff
@@ -12,26 +12,50 @@
...
+
+    ## ------- Set up our list of numbers -------
+    S = [25, 7, 13, 31, 42, 17, 21, 10]
+
+    # TODO: Enter your token here
+    token = 'Your-Token-Here'
+
+    ## ------- Set up our QUBO dictionary -------
+
+    Q = define_QUBO(S)
```
Comment (m3ller) [inline npp.py]: I still think this line should be `Q = get_qubo(S)`.

However, I would understand if what I am suggesting does not make sense in a classroom context
